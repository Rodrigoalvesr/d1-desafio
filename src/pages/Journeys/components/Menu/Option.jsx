import React from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import Icon from '@Components/Icon';
import Badge from '@Components/Badge';
import getStatus from '@/common/getStatus';
import { makeStyles } from '@material-ui/core/styles';
import classnames from 'classnames';

const useStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
    marginBottom: 10,
    userSelect: 'none',
  },
  content: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: '100%',
    cursor: 'pointer',
    '&:hover': {
      '& $name': {
        color: theme.palette.primary.main,
      },
      '& $badge:not(.selected)': {
        background: '#EBEEF6',
      },
    },
  },
  icon: {
    width: 16,
    height: 16,
    marginRight: 10,
  },
  badge: {},
  name: {
    fontSize: 13,
    '&.selected': {
      fontWeight: 'bold',
      color: '#117EFF',
    },
  },
}));

function Option({ id, name, quantity, selected, onClick }) {
  const classes = useStyles();
  const status = React.useMemo(() => getStatus(id), [id]);

  return (
    <div className={classes.root}>
      <Icon className={classes.icon} name={status.icon} color={status.color} />

      <div className={classes.content} onClick={onClick}>
        <Typography className={classnames(classes.name, { selected })}>
          {name}
        </Typography>

        <Badge selected={selected} value={quantity} className={classes.badge} />
      </div>
    </div>
  );
}

Option.defaultProps = {
  selected: false,
};

Option.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  quantity: PropTypes.number.isRequired,
  onClick: PropTypes.func.isRequired,
  selected: PropTypes.bool,
};

export default Option;
