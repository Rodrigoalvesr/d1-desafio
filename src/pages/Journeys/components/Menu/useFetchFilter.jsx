import { useQuery } from 'react-query';
import api from '@/services/api';

function useFetchFilter() {
  return useQuery('filter', () => api.get('/filter').then(({ data }) => data));
}

export default useFetchFilter;
