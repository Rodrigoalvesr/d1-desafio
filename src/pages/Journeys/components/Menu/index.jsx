import React from 'react';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Option from './Option';
import { useSelector, useDispatch } from 'react-redux';
import useFetchFilter from './useFetchFilter';

const useStyles = makeStyles({
  root: {
    display: 'flex',
    flexDirection: 'column',
    marginLeft: 90,
    width: 190,
  },
  title: {
    color: '#3E4157',
    fontWeight: 'bold',
    fontSize: 16,
    marginBottom: 30,
  },
});

function Menu() {
  const classes = useStyles();
  const dispatch = useDispatch();
  const { filter } = useSelector((journey) => journey);

  const { data = [] } = useFetchFilter();

  const handleChangeFilter = React.useCallback(
    (id) => () => {
      dispatch({
        type: 'JOURNEY_SET_FILTER',
        payload: id,
      });
    },
    [dispatch]
  );

  return (
    <div className={classes.root}>
      <Typography className={classes.title}>Jornadas</Typography>
      {data.map((item) => (
        <Option
          key={`filter_${item.id}`}
          id={item.id}
          name={item.name}
          quantity={item.quantity}
          selected={item.id === filter}
          onClick={handleChangeFilter(item.id)}
        />
      ))}
    </div>
  );
}

export default Menu;
