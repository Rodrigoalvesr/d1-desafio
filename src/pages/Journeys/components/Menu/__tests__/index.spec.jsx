import React from 'react';
import renderer from 'react-test-renderer';
import Menu from '../index';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import useFetchFilter from '../useFetchFilter';

const mockStore = configureMockStore([thunk]);
const store = mockStore({
  filter: 0,
});
jest.mock('../useFetchFilter', () => jest.fn());

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useCallback: (cb) => () => jest.fn(cb()()),
}));

describe('Menu', () => {
  it('Renders correctly', () => {
    useFetchFilter.mockReturnValue({
      data: [
        {
          id: 0,
          name: 'Todos',
          quantity: 8,
        },
        {
          id: 1,
          name: 'Em execução',
          quantity: 2,
        },
      ],
    });
    const component = renderer
      .create(
        <Provider store={store}>
          <Menu />
        </Provider>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
