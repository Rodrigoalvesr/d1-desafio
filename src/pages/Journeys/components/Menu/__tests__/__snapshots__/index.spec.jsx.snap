// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Renders correctly 1`] = `
<div
  className="makeStyles-root-1"
>
  <p
    className="MuiTypography-root makeStyles-title-2 MuiTypography-body1"
  >
    Jornadas
  </p>
  <div
    className="makeStyles-root-3"
  >
    <svg
      className="makeStyles-root-8 makeStyles-root-9 makeStyles-icon-5"
    >
      table.svg
    </svg>
    <div
      className="makeStyles-content-4"
      onClick={[MockFunction]}
    >
      <p
        className="MuiTypography-root makeStyles-name-7 selected MuiTypography-body1"
      >
        Todos
      </p>
      <div
        className="makeStyles-root-10 makeStyles-badge-6 selected"
      >
        <p
          className="MuiTypography-root makeStyles-text-11 selected MuiTypography-body1"
        >
          8
        </p>
      </div>
    </div>
  </div>
  <div
    className="makeStyles-root-3"
  >
    <svg
      className="makeStyles-root-8 makeStyles-root-12 makeStyles-icon-5"
    >
      paper-plane.svg
    </svg>
    <div
      className="makeStyles-content-4"
      onClick={[MockFunction]}
    >
      <p
        className="MuiTypography-root makeStyles-name-7 MuiTypography-body1"
      >
        Em execução
      </p>
      <div
        className="makeStyles-root-10 makeStyles-badge-6"
      >
        <p
          className="MuiTypography-root makeStyles-text-11 MuiTypography-body1"
        >
          2
        </p>
      </div>
    </div>
  </div>
</div>
`;
